components/brave_rewards/core/endpoints/brave/post_challenges.cc

```diff
@@ -0,0 +1,106 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_rewards/core/endpoints/brave/post_challenges.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "brave/components/brave_rewards/core/common/callback_helpers.h"
+#include "brave/components/brave_rewards/core/common/request_signer.h"
+#include "brave/components/brave_rewards/core/common/url_loader.h"
+#include "brave/components/brave_rewards/core/endpoint/promotion/promotions_util.h"
+#include "brave/components/brave_rewards/core/wallet/wallet.h"
+#include "net/http/http_status_code.h"
+
+namespace brave_rewards::internal::endpoints {
+
+PostChallenges::PostChallenges(RewardsEngineImpl& engine)
+    : RewardsEngineHelper(engine) {}
+
+PostChallenges::~PostChallenges() = default;
+
+void PostChallenges::Request(RequestCallback callback) {
+  auto request = CreateRequest();
+  if (!request) {
+    DeferCallback(FROM_HERE, std::move(callback),
+                  base::unexpected(Error::kInvalidRequest));
+    return;
+  }
+
+  Get<URLLoader>().Load(
+      std::move(request), URLLoader::LogLevel::kBasic,
+      base::BindOnce(&PostChallenges::OnResponse, weak_factory_.GetWeakPtr(),
+                     std::move(callback)));
+}
+
+mojom::UrlRequestPtr PostChallenges::CreateRequest() {
```

Comment:
This is trusting the JSON to be parsed before santizing it which is considered unsafe. A safer way to make this request and parse the response would be with [api_request_helper](https://github.com/brave/brave-core/blob/master/components/api_request_helper/api_request_helper.cc) Can we move to using this instead so that we aren't redoing the sanitization logic implemented there with [data_decoder](https://github.com/brave/brave-core/blob/master/components/api_request_helper/api_request_helper.cc#L355C3-L359)

Reply:
For Rewards engine code, all network requests go through [here](https://github.com/brave/brave-core/blob/c37a456053db19cb19e8876b228a7240777bb636/components/brave_rewards/browser/rewards_service_impl.cc#L1048) which I believe was added to ensure that all JSON responses are sanitized.

I think we should improve this setup in the future, though, so that we don't need to use `JSONReader` to parse JSON at al.

Reply:
Just went through this and can confirm we're using data_decoder properly. Thanks for sending the links through for this and I'll add it to my notes that Rewards is doing this sanitization via the `URLLoader.Load` call.